order/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+//! Contains the order type as described by the specification with serialization as described by the openapi documentation.
+//!
+//! This is in its own crate because we want to share this module between the orderbook and the solver.
+
+use chrono::{offset::Utc, DateTime};
+use primitive_types::H160;
+use serde::{de, Deserialize, Serialize};
+use std::fmt;
+
+#[derive(Eq, PartialEq, Clone, Copy, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub enum FillType {
+    FillKill,
+    Partial,
+}
+
+#[derive(Eq, PartialEq, Clone, Copy, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub enum OrderType {
+    Buy,
+    Sell,
+}
+
+#[derive(Eq, PartialEq, Clone, Copy, Debug)]
+pub struct Signature {
+    pub v: u8,
+    pub r: [u8; 32],
+    pub s: [u8; 32],
+}
+
+/// An order as provided to the orderbook by the frontend.
+#[derive(Eq, PartialEq, Clone, Copy, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct UserProvided {
+    #[serde(with = "h160_hex")]
+    pub buy_token: H160,
+    #[serde(with = "serde_with::rust::display_fromstr")]
+    pub buy_amount: u128,
+    #[serde(with = "h160_hex")]
+    pub sell_token: H160,
+    #[serde(with = "serde_with::rust::display_fromstr")]
+    pub sell_amount: u128,
+    #[serde(with = "serde_with::rust::display_fromstr")]
+    pub sell_token_tip: u128,
+    pub order_type: OrderType,
+    pub fill_type: FillType,
+    pub nonce: u32,
+    pub valid_to: u32,
+    pub signature: Signature,
+}
+
+/// An order that is returned when querying the orderbook.
+///
+/// Contains extra fields thats are populated by the orderbook.
+#[derive(Eq, PartialEq, Clone, Copy, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Full {
+    pub creation_time: DateTime<Utc>,
+    #[serde(with = "h160_hex")]
+    pub owner: H160,
+    #[serde(flatten)]
+    pub user_provided: UserProvided,
+}
+
+impl Serialize for Signature {
```

Comment:
Maybe signature could be its own model?

Reply:
What do you mean by it's own model? It is already it's own struct.

Reply:
sorry, I meant module and this its own file.

Reply:
We will see how the file grows. Not yet needed.